File: src/day04.rs
use std::error::Error;

type Range = (usize, usize);
type WorkerPair = (Range, Range);

fn parse(input: &str) -> Vec<WorkerPair> {
    input
        .lines()
        .map(|line| {
            let pair: Vec<Range> = line
                .split(',')
                .map(|part| {
                    let xs: Vec<usize> = part
                        .split('-')
                        .map(|s| s.to_string().parse::<usize>().unwrap())
                        .collect();
                    (xs[0], xs[1])
                })
                .collect();
            (pair[0], pair[1])
        })
        .collect()
}

fn fully_contained(wp: WorkerPair) -> bool {
    let ((a, b), (x, y)) = wp;
    (a <= x && b >= y) || (x <= a && y >= b)
}

fn overlap(wp: WorkerPair) -> bool {
    let ((a, b), (x, y)) = wp;
    (a <= y && a >= x) || (b <= y && b >= x) || (x <= b && x >= a) || (y <= b && y >= a)
}

fn main() -> Result<(), Box<dyn Error>> {
    let input = include_str!("../input/day04.txt");
    let elf_pairs = parse(input);
    let contained_count: usize =
        elf_pairs.iter().fold(
            0,
            |acc, wp| {
                if fully_contained(*wp) {
                    acc + 1
                } else {
                    acc
                }
            },
        );

    let elf_pairs = parse(input);
    let overlap_count: usize = elf_pairs
        .iter()
        .fold(0, |acc, wp| if overlap(*wp) { acc + 1 } else { acc });
    println!("{:?}, {:?}", contained_count, overlap_count);
    Ok(())
}

File: ruby/src/day04.rb
def parse(input)
  input.lines.map do |line|
    line.split(',').map do |part|
      part.split('-').map(&:to_i)
    end
  end
end

def fully_contained(wp)
  ((a, b), (x, y)) = wp
  (a <= x && b >= y) || (x <= a && y >= b)
end

def overlap(wp)
  ((a, b), (x, y)) = wp
  (a <= y && a >= x) || (b <= y && b >= x) || (x <= b && x >= a) || (y <= b && y >= a)
end

def contained_count(worker_pairs)
  worker_pairs.count { |wp| fully_contained(wp) }
end

def overlap_count(worker_pairs)
  worker_pairs.count { |wp| overlap(wp) }
end

if $PROGRAM_NAME == __FILE__
  input = File.read('../input/day04.txt')

  elf_pairs = parse(input)

  part_one = contained_count(elf_pairs)
  puts "Contained count (part 1): #{part_one}"

  part_two = overlap_count(elf_pairs)
  puts "Overlap Count (part 2): #{part_two}"
end
